plugin/ipam/cni.go

```diff
@@ -35,31 +36,80 @@ func (i *Ipam) Allocate(args *skel.CmdArgs) (types.Result, error) {
 	if containerID == "" {
 		return nil, fmt.Errorf("Weave CNI Allocate: blank container name")
 	}
-	var ipnet *net.IPNet
 
-	if conf.Subnet == "" {
-		ipnet, err = i.weave.AllocateIP(containerID, false)
+	var ipconfigs []*current.IPConfig
+
+	if len(conf.IPs) > 0 {
+		// configuration includes desired IPs
+
+		var ips []net.IP
+		for _, ip := range conf.IPs {
+			ip4 := net.ParseIP(ip).To4()
+			ip16 := net.ParseIP(ip).To16()
+
+			if ip4 == nil && ip16 == nil {
+				return nil, errors.New("provided value is not an IP")
+			}
+
+			if ip4 == nil && ip16 != nil {
+				return nil, errors.New("allocation of ipv6 addresses is not implemented")
+			}
+
+			ips = append(ips, ip4)
+		}
+
+		for j := range ips {
+			ipnet := &net.IPNet{
+				IP:   ips[j],
+				Mask: ips[j].DefaultMask(),
```

Comment:
This seems wrong; the IPs are coming back as `10.32.1.42/8` rather than `10.32.1.42/12` which would match the default subnet used by Weave Net.